### PR TITLE
test: Improve test coverage for device handlers

### PR DESCRIPTION
**Pull request by SantoshMadhukar-K**
Add comprehensive test coverage for the device handler modules under src/agent/src/device, including matcher behavior, edge cases, network, nvdimm, scsi, and vfio device paths.

```
cargo test device
warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /root/temp_repos/myforks/kata-containers/src/agent/Cargo.toml
workspace: /root/temp_repos/myforks/kata-containers/Cargo.toml
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.90s
     Running unittests src/main.rs (/root/temp_repos/myforks/kata-containers/target/debug/deps/kata_agent-86dff20ddd35f1e6)

running 108 tests
test device::block_device_handler::tests::test_virtio_blk_mmio_matcher_basic_matching::case_2_vdb_matches_vdb ... ok
test device::block_device_handler::tests::test_virtio_blk_mmio_matcher_basic_matching::case_1_vda_matches_vda ... ok
test device::block_device_handler::tests::test_virtio_blk_mmio_matcher_basic_matching::case_4_vdb_rejects_vda ... ok
test device::block_device_handler::tests::test_virtio_blk_mmio_matcher_empty_devname ... ok
test device::block_device_handler::tests::test_virtio_blk_mmio_matcher_wrong_device_suffix ... ok
test device::block_device_handler::tests::test_virtio_blk_mmio_matcher_wrong_subsystem::case_1_wrong_subsystem ... ok
test device::block_device_handler::tests::test_virtio_blk_mmio_matcher_basic_matching::case_3_vda_rejects_vdb ... ok
test device::block_device_handler::tests::test_virtio_blk_pci_matcher_accepts_any_virtio_number::case_1_virtio4 ... ok
test device::block_device_handler::tests::test_virtio_blk_pci_matcher_accepts_any_virtio_number::case_2_virtio99 ... ok
test device::block_device_handler::tests::test_virtio_blk_pci_matcher_basic_matching::case_1_matcher_a_matches_uev_a ... ok
test device::block_device_handler::tests::test_virtio_blk_pci_matcher_accepts_any_virtio_number::case_3_virtio0 ... ok
test device::block_device_handler::tests::test_virtio_blk_pci_matcher_basic_matching::case_3_matcher_a_rejects_uev_b ... ok
test device::block_device_handler::tests::test_virtio_blk_pci_matcher_basic_matching::case_2_matcher_b_matches_uev_b ... ok
test device::block_device_handler::tests::test_virtio_blk_pci_matcher_correct_match ... ok
test device::block_device_handler::tests::test_virtio_blk_pci_matcher_basic_matching::case_4_matcher_b_rejects_uev_a ... ok
test device::block_device_handler::tests::test_virtio_blk_pci_matcher_empty_devname ... ok
test device::block_device_handler::tests::test_virtio_blk_pci_matcher_missing_virtio_component ... ok
test device::block_device_handler::tests::test_virtio_blk_pci_matcher_rejects_partitions::case_1_partition_vda1 ... ok
test device::block_device_handler::tests::test_virtio_blk_pci_matcher_rejects_partitions::case_2_partition_vda91 ... ok
test device::block_device_handler::tests::test_virtio_blk_pci_matcher_wrong_subsystem::case_1_wrong_subsystem ... ok
test device::network_device_handler::tests::test_net_pci_matcher_basic_matching::case_1_matcher_a_matches_uev_a ... ok
test device::network_device_handler::tests::test_net_pci_matcher_basic_matching::case_3_matcher_a_rejects_uev_b ... ok
test device::network_device_handler::tests::test_net_pci_matcher_basic_matching::case_4_matcher_b_rejects_uev_a ... ok
test device::network_device_handler::tests::test_net_pci_matcher_empty_interface ... ok
test device::network_device_handler::tests::test_net_pci_matcher_with_net_substring ... ok
test device::network_device_handler::tests::test_net_pci_matcher_wrong_action::case_1_wrong_action ... ok
test device::network_device_handler::tests::test_net_pci_matcher_wrong_devpath ... ok
test device::network_device_handler::tests::test_net_pci_matcher_wrong_subsystem::case_1_wrong_subsystem ... ok
test device::nvdimm_device_handler::tests::test_pmem_block_matcher_basic_matching::case_1_pmem0_matches_pmem0 ... ok
test device::nvdimm_device_handler::tests::test_pmem_block_matcher_basic_matching::case_2_pmem1_matches_pmem1 ... ok
test device::network_device_handler::tests::test_net_pci_matcher_basic_matching::case_2_matcher_b_matches_uev_b ... ok
test device::nvdimm_device_handler::tests::test_pmem_block_matcher_basic_matching::case_3_pmem0_rejects_pmem1 ... ok
test device::nvdimm_device_handler::tests::test_pmem_block_matcher_basic_matching::case_4_pmem1_rejects_pmem0 ... ok
test device::nvdimm_device_handler::tests::test_pmem_block_matcher_empty_devname ... ok
test device::nvdimm_device_handler::tests::test_pmem_block_matcher_wrong_subsystem::case_1_wrong_subsystem ... ok
test device::nvdimm_device_handler::tests::test_pmem_block_matcher_wrong_suffix ... ok
test device::scsi_device_handler::tests::test_scsi_block_matcher_basic_matching::case_1_addr_a_matches_uev_a ... ok
test device::scsi_device_handler::tests::test_scsi_block_matcher_basic_matching::case_2_addr_b_matches_uev_b ... ok
test device::scsi_device_handler::tests::test_scsi_block_matcher_basic_matching::case_3_addr_a_rejects_uev_b ... ok
test device::nvdimm_device_handler::tests::test_pmem_block_matcher_wrong_prefix ... ok
test device::scsi_device_handler::tests::test_scsi_block_matcher_basic_matching::case_4_addr_b_rejects_uev_a ... ok
test device::scsi_device_handler::tests::test_scsi_block_matcher_different_addresses::case_2_addr_0_0_rejects ... ok
test device::scsi_device_handler::tests::test_scsi_block_matcher_different_addresses::case_1_addr_1_1_matches ... ok
test device::scsi_device_handler::tests::test_scsi_block_matcher_empty_devname ... ok
test device::scsi_device_handler::tests::test_scsi_block_matcher_rejects_partitions::case_1_whole_disk ... ok
test device::scsi_device_handler::tests::test_scsi_block_matcher_rejects_partitions::case_2_partition ... ok
test device::scsi_device_handler::tests::test_scsi_block_matcher_wrong_path ... ok
test device::scsi_device_handler::tests::test_scsi_block_matcher_wrong_subsystem::case_1_wrong_subsystem ... ok
test device::tests::test_bus_of_addr::case_1_normal ... ok
test device::tests::test_bus_of_addr::case_2_max_values ... ok
test device::tests::test_bus_of_addr::case_3_all_zeros ... ok
test device::tests::test_dev_update_new ... ok
test device::tests::test_device_info_new::case_1_char_device ... ok
test device::tests::test_device_info_new::case_2_block_device ... ok
test device::tests::test_device_info_new::case_3_nonexistent ... ok
test device::tests::test_device_info_new::case_4_empty_path ... ok
test device::tests::test_handle_cdi_devices ... ok
test device::tests::test_infer_bus_from_child_devices ... ok
test device::tests::test_online_device ... ok
test device::tests::test_get_device_name ... ok
test device::tests::test_parse_pci_bdf_name::case_1_valid_zeros ... ok
test device::tests::test_parse_pci_bdf_name::case_2_valid_normal ... ok
test device::tests::test_parse_pci_bdf_name::case_3_valid_max ... ok
test device::tests::test_parse_pci_bdf_name::case_4_invalid_text ... ok
test device::tests::test_parse_pci_bdf_name::case_5_empty_string ... ok
test device::tests::test_parse_pci_bdf_name::case_6_invalid_format ... ok
test device::tests::test_parse_pci_bdf_name::case_7_random_string ... ok
test device::tests::test_pcipath_to_sysfs ... ok
test device::tests::test_pcipath_to_sysfs_fallback_child_device_scan ... ok
test device::tests::test_read_single_bus_from_pci_bus_dir ... ok
test device::tests::test_spec_update_conversions ... ok
test device::tests::test_unique_bus_from_pci_addresses::case_1_single_bus ... ok
test device::tests::test_unique_bus_from_pci_addresses::case_2_multiple_buses ... ok
test device::tests::test_unique_bus_from_pci_addresses::case_3_empty_list ... ok
test device::tests::test_update_device_cgroup ... ok
test device::tests::test_update_env_pci ... ok
test device::tests::test_update_env_pci_missing_cid_in_pcimap ... ok
test device::tests::test_update_env_pci_non_pci_addresses ... ok
test device::tests::test_update_env_pci_unknown_host_bdf ... ok
test device::tests::test_update_spec_devices ... ok
test device::tests::test_update_spec_devices_char_block_conflict ... ok
test device::tests::test_update_spec_devices_final_path ... ok
test device::tests::test_update_spec_devices_guest_host_conflict ... ok
test device::vfio_device_handler::tests::test_pci_iommu_group ... ok
test device::vfio_device_handler::tests::test_pci_matcher_basic_matching::case_1_relpath_a_matches ... ok
test device::vfio_device_handler::tests::test_pci_matcher_basic_matching::case_2_relpath_b_matches ... ok
test device::vfio_device_handler::tests::test_pci_matcher_basic_matching::case_3_relpath_a_rejects_b ... ok
test device::vfio_device_handler::tests::test_pci_matcher_basic_matching::case_4_relpath_b_rejects_a ... ok
test device::vfio_device_handler::tests::test_pci_matcher_different_root_complex ... ok
test device::vfio_device_handler::tests::test_pci_matcher_partial_path ... ok
test device::vfio_device_handler::tests::test_split_vfio_pci_option::case_1_valid_option ... ok
test device::vfio_device_handler::tests::test_split_vfio_pci_option::case_2_too_many_equals ... ok
test device::vfio_device_handler::tests::test_pci_driver_override ... ok
test device::vfio_device_handler::tests::test_vfio_matcher_basic_matching::case_1_group_1_matches ... ok
test device::vfio_device_handler::tests::test_split_vfio_pci_option::case_3_missing_equals ... ok
test device::vfio_device_handler::tests::test_vfio_matcher_basic_matching::case_2_group_22_matches ... ok
test device::vfio_device_handler::tests::test_vfio_matcher_basic_matching::case_3_group_1_rejects_22 ... ok
test device::vfio_device_handler::tests::test_vfio_matcher_basic_matching::case_4_group_22_rejects_1 ... ok
test device::vfio_device_handler::tests::test_vfio_matcher_partial_match ... ok
test device::vfio_device_handler::tests::test_vfio_matcher_wrong_devpath ... ok
test storage::multi_layer_erofs::tests::test_get_partition_device_path::case_1 ... ok
test storage::multi_layer_erofs::tests::test_get_partition_device_path::case_2 ... ok
test storage::multi_layer_erofs::tests::test_get_partition_device_path::case_3 ... ok
test storage::multi_layer_erofs::tests::test_get_partition_device_path::case_4 ... ok
test storage::multi_layer_erofs::tests::test_get_partition_device_path::case_5 ... ok
test storage::multi_layer_erofs::tests::test_get_partition_device_path::case_6 ... ok
test storage::multi_layer_erofs::tests::test_get_partition_device_path::case_7 ... ok
test storage::multi_layer_erofs::tests::test_get_partition_device_path::case_8 ... ok

test result: ok. 108 passed; 0 failed; 0 ignored; 0 measured; 232 filtered out; finished in 0.12s
```

### CODE REVIEW

**Comment by stevenhorsman** [comment]
@SantoshMadhukar-K - there is a merge conflict here. Can you rebase and resolve it please.

**Comment by SantoshMadhukar-K** [comment]
> @SantoshMadhukar-K - there is a merge conflict here. Can you rebase and resolve it please.

@stevenhorsman Done Rebase. Please Review. Thanks

File: src/agent/src/device/block_device_handler.rs
Comment (stevenhorsman) [inline src/agent/src/device/block_device_handler.rs]: Did you consider using rstest to split this down as it seems to be testing multiple things in a single test e.g.
```
use rstest::rstest;

// Helper to create a standard uevent for testing
fn create_mmio_uevent(devname: &str, mmio_id: u32, virtio_id: u32) -> crate::uevent::Uevent {
    let mut uev = crate::uevent::Uevent::default();
    uev.action = crate::linux_abi::U_EVENT_ACTION_ADD.to_string();
    uev.subsystem = BLOCK.to_string();
    uev.devname = devname.to_string();
    uev.devpath = format!(
        "/sys/devices/virtio-mmio-cmdline/virtio-mmio.{}/virtio{}/block/{}",
        mmio_id, virtio_id, devname
    );
    uev
}

#[rstest]
#[case::vda_matches_vda("vda", "vda", 0, 0, true)]
#[case::vdb_matches_vdb("vdb", "vdb", 4, 4, true)]
#[case::vda_rejects_vdb("vda", "vdb", 0, 0, false)]
#[case::vdb_rejects_vda("vdb", "vda", 4, 4, false)]
#[tokio::test]
async fn test_virtio_blk_mmio_matcher_basic_matching(
    #[case] matcher_devname: &str,
    #[case] uevent_devname: &str,
    #[case] mmio_id: u32,
    #[case] virtio_id: u32,
    #[case] should_match: bool,
) {
    let matcher = VirtioBlkMmioMatcher::new(matcher_devname);
    let uev = create_mmio_uevent(uevent_devname, mmio_id, virtio_id);
    
    assert_eq!(
        matcher.is_match(&uev),
        should_match,
        "Matcher for '{}' should {} uevent for '{}'",
        matcher_devname,
        if should_match { "match" } else { "reject" },
        uevent_devname
    );
}

#[rstest]
#[case::wrong_subsystem(test_helpers::SUBSYSTEM_NET, "Wrong subsystem should be rejected")]
#[tokio::test]
async fn test_virtio_blk_mmio_matcher_wrong_subsystem(
    #[case] wrong_subsystem: &str,
    #[case] description: &str,
) {
    let devname = "vda";
    let matcher = VirtioBlkMmioMatcher::new(devname);
    let mut uev = create_mmio_uevent(devname, 0, 0);
    uev.subsystem = wrong_subsystem.to_string();
    
    assert!(!matcher.is_match(&uev), "{}", description);
}

#[tokio::test]
async fn test_virtio_blk_mmio_matcher_empty_devname() {
    let devname = "vda";
    let matcher = VirtioBlkMmioMatcher::new(devname);
    let mut uev = create_mmio_uevent(devname, 0, 0);
    uev.devname = String::new();
    
    assert!(
        !matcher.is_match(&uev),
        "Matcher should reject uevent with empty devname"
    );
}

#[rstest]
#[case::wrong_device_suffix("vda", "vdc", "Matcher should reject uevent with wrong device suffix")]
#[case::different_mmio_id("vda", "vda", "Matcher should reject uevent with different MMIO ID")]
#[tokio::test]
async fn test_virtio_blk_mmio_matcher_device_specific_edge_cases(
    #[case] matcher_devname: &str,
    #[case] uevent_devname: &str,
    #[case] description: &str,
) {
    let matcher = VirtioBlkMmioMatcher::new(matcher_devname);
    let mut uev = create_mmio_uevent(matcher_devname, 0, 0);
    // Modify to create the wrong suffix scenario
    uev.devpath = format!(
        "/sys/devices/virtio-mmio-cmdline/virtio-mmio.0/virtio0/block/{}",
        uevent_devname
    );
    
    assert!(!matcher.is_match(&uev), "{}", description);
}
```
?

File: src/agent/src/device/mod.rs [hunk 1]
```diff
@@ -948,6 +1034,187 @@ mod tests {
...
+    fn test_bus_of_addr() {
+        let addr1 = pci::Address::new(0, 1, pci::SlotFn::new(2, 3).unwrap());
+        assert_eq!(bus_of_addr(&addr1).unwrap(), "0000:01");
+
+        let addr2 = pci::Address::new(0xffff, 0xff, pci::SlotFn::new(0x1f, 7).unwrap());
+        assert_eq!(bus_of_addr(&addr2).unwrap(), "ffff:ff");
+
+        let addr3 = pci::Address::new(0, 0, pci::SlotFn::new(0, 0).unwrap());
+        assert_eq!(bus_of_addr(&addr3).unwrap(), "0000:00");
+    }
```
Comment (stevenhorsman) [inline src/agent/src/device/mod.rs:1155]: Same here:
```
#[rstest]
#[case::normal(0, 1, 2, 3, "0000:01")]
#[case::max_values(0xffff, 0xff, 0x1f, 7, "ffff:ff")]
#[case::all_zeros(0, 0, 0, 0, "0000:00")]
#[test]
fn test_bus_of_addr(
    #[case] domain: u16,
    #[case] bus: u8,
    #[case] slot: u8,
    #[case] func: u8,
    #[case] expected: &str,
) {
    let addr = pci::Address::new(domain, bus, pci::SlotFn::new(slot, func).unwrap());
    assert_eq!(bus_of_addr(&addr).unwrap(), expected);
}
```

File: src/agent/src/device/mod.rs [hunk 2]
```diff
@@ -948,6 +1034,187 @@ mod tests {
...
+
+    #[test]
+    fn test_device_info_new_errors() {
+        let result = DeviceInfo::new("/nonexistent/path", true);
+        assert!(result.is_err());
+
+        let result_empty = DeviceInfo::new("", true);
+        assert!(result_empty.is_err());
+    }
```
Comment (stevenhorsman) [inline src/agent/src/device/mod.rs:1308]: Same here:
```
#[rstest]
#[case::char_device("/dev/null", true, "c", true)]
#[case::block_device("/", false, "b", true)]
#[case::nonexistent("/nonexistent/path", true, "", false)]
#[case::empty_path("", true, "", false)]
#[test]
fn test_device_info_new(
    #[case] path: &str,
    #[case] is_char: bool,
    #[case] expected_type: &str,
    #[case] should_succeed: bool,
) {
    let result = DeviceInfo::new(path, is_char);
    
    if should_succeed {
        let info = result.unwrap();
        assert_eq!(info.cgroup_type, expected_type);
        assert!(info.guest_major >= 0);
        assert!(info.guest_minor >= 0);
    } else {
        assert!(result.is_err());
    }
}
```

File: src/agent/src/device/mod.rs [hunk 3]
```diff
@@ -948,6 +1034,187 @@ mod tests {
...
+        let addrs_multi = vec![
+            pci::Address::new(0, 1, pci::SlotFn::new(0, 0).unwrap()),
+            pci::Address::new(0, 2, pci::SlotFn::new(0, 0).unwrap()),
+        ];
+        assert!(unique_bus_from_pci_addresses(&addrs_multi).is_err());
+
+        // Empty list
+        let addrs_empty: Vec<pci::Address> = vec![];
+        assert!(unique_bus_from_pci_addresses(&addrs_empty).is_err());
+    }
```
Comment (stevenhorsman) [inline src/agent/src/device/mod.rs]: And here:
```
#[rstest]
#[case::single_bus(
    vec![(0, 1, 0, 0), (0, 1, 1, 0), (0, 1, 2, 0)],
    Some("0000:01")
)]
#[case::multiple_buses(
    vec![(0, 1, 0, 0), (0, 2, 0, 0)],
    None
)]
#[case::empty_list(vec![], None)]
#[test]
fn test_unique_bus_from_pci_addresses(
    #[case] addr_tuples: Vec<(u16, u8, u8, u8)>,
    #[case] expected: Option<&str>,
) {
    let addrs: Vec<pci::Address> = addr_tuples
        .into_iter()
        .map(|(d, b, s, f)| pci::Address::new(d, b, pci::SlotFn::new(s, f).unwrap()))
        .collect();
    
    match expected {
        Some(bus) => assert_eq!(unique_bus_from_pci_addresses(&addrs).unwrap(), bus),
        None => assert!(unique_bus_from_pci_addresses(&addrs).is_err()),
    }
}
```

File: src/agent/src/device/block_device_handler.rs
Comment (stevenhorsman) [reply src/agent/src/device/block_device_handler.rs]: I think this applies to the other handler matcher test as well?

**Review by stevenhorsman — COMMENTED** [review]
Some test improvement suggestions

File: src/agent/src/device/block_device_handler.rs
Comment (SantoshMadhukar-K) [reply src/agent/src/device/block_device_handler.rs]: I have checked all the files to use the rstest with bob. Once again i will check. thanks

**Comment by stevenhorsman** [comment]
Hi Santosh this needs rebasing, can you fix the trailing whitespace and run make check before submitting the PR to address the cargo fmt and clippy issues

**Comment by SantoshMadhukar-K** [comment]
> Hi Santosh this needs rebasing, can you fix the trailing whitespace and run make check before submitting the PR to address the cargo fmt and clippy issues

Sure steven I have done checks but not sure where it comming from. Sure will check these before subbmitting to PR. Thanks

**Comment by SantoshMadhukar-K** [comment]
@stevenhorsman I have updated the code with all error fixes, fmt and clippy issues. Please review. Thanks

**Review by stevenhorsman — APPROVED** [review]
Thanks for the improvements, I think this is probably as good as we will get. Can you comment on the co-pilot raised issues; either that you've addressed them and the resolve them, or why you don't think they are valid. Thanks

**Comment by SantoshMadhukar-K** [comment]
> Thanks for the improvements, I think this is probably as good as we will get. Can you comment on the co-pilot raised issues; either that you've addressed them and the resolve them, or why you don't think they are valid. Thanks

Sure @stevenhorsman I already resolved them. I will address the comments. Thanks

File: src/agent/src/device/mod.rs [hunk 4]
```diff
@@ -732,6 +732,93 @@ pub fn online_device(path: &str) -> Result<()> {
...
+    ) {
+        assert!(
+            matcher_a.is_match(uev_a),
+            "Matcher A should match its own uevent A"
+        );
+        assert!(
+            matcher_b.is_match(uev_b),
+            "Matcher B should match its own uevent B"
+        );
+    }
```
[the remark this answers is not in the document]
Comment (SantoshMadhukar-K) [reply src/agent/src/device/mod.rs]: Resolved this issue. Thanks

File: src/agent/src/device/nvdimm_device_handler.rs [hunk 5]
```diff
@@ -81,3 +81,50 @@ impl UeventMatcher for PmemBlockMatcher {
...
+        uev_a.devpath = format!("{}/LNXSYSTM:00/LNXSYBUS:00/ACPI0012:00/ndbus0/region0/btt0.0/block/{}", ACPI_DEV_PATH, devname_a);
+        let matcher_a = PmemBlockMatcher::new(devname_a);
+
+        let mut uev_b = uev_a.clone();
+        uev_b.devname = devname_b.to_string();
+        uev_b.devpath = format!("{}/LNXSYSTM:00/LNXSYBUS:00/ACPI0012:00/ndbus0/region1/btt1.0/block/{}", ACPI_DEV_PATH, devname_b);
+        let matcher_b = PmemBlockMatcher::new(devname_b);
+
+        // Test correct matches and cross-matching
+        test_helpers::assert_matchers_match_own(&matcher_a, &matcher_b, &uev_a, &uev_b);
```
[the remark this answers is not in the document]
Comment (SantoshMadhukar-K) [reply src/agent/src/device/nvdimm_device_handler.rs]: Added the new negative cases. Thanks

File: src/agent/src/device/scsi_device_handler.rs [hunk 6]
```diff
@@ -128,9 +129,33 @@ mod tests {
             format!("{root_bus}/0000:00:00.0/virtio0/host0/target0:0:2/0:0:{addr_b}/block/sdb");
         let matcher_b = ScsiBlockMatcher::new(addr_b);
 
-        assert!(matcher_a.is_match(&uev_a));
-        assert!(matcher_b.is_match(&uev_b));
-        assert!(!matcher_b.is_match(&uev_a));
-        assert!(!matcher_a.is_match(&uev_b));
+        // Test correct matches and cross-matching
+        test_helpers::assert_matchers_match_own(&matcher_a, &matcher_b, &uev_a, &uev_b);
```
[the remark this answers is not in the document]
Comment (SantoshMadhukar-K) [reply src/agent/src/device/scsi_device_handler.rs]: Added the "no cross match" testcase. Thanks

File: src/agent/src/device/vfio_device_handler.rs [hunk 7]
```diff
@@ -474,10 +475,24 @@ mod tests {
         uev_b.devpath = format!("/devices/virtual/vfio/{grpb}");
         let matcher_b = VfioMatcher::new(grpb);
 
-        assert!(matcher_a.is_match(&uev_a));
-        assert!(matcher_b.is_match(&uev_b));
-        assert!(!matcher_b.is_match(&uev_a));
-        assert!(!matcher_a.is_match(&uev_b));
+        // Test correct matches and cross-matching
+        test_helpers::assert_matchers_match_own(&matcher_a, &matcher_b, &uev_a, &uev_b);
```
[the remark this answers is not in the document]
Comment (SantoshMadhukar-K) [reply src/agent/src/device/vfio_device_handler.rs]: Added the negative cross match assertion testcases. Thanks

**Comment by BbolroC** [comment]
@SantoshMadhukar-K Please check out https://github.com/kata-containers/kata-containers/actions/runs/27744355094/job/82079186444?pr=13078 and fix the errors. Thanks!

**Comment by stevenhorsman** [comment]
@SantoshMadhukar-K - can you rebase please as your last push is still 29 commits behind main. Thanks

**Review by BbolroC — APPROVED** [review]
